terraform/full_environment/events_ingestion.tf

```diff
@@ -0,0 +1,188 @@
+    "bigquery.jobs.create"
+  ]
+}
+
+# binding role to service account
+resource "google_project_iam_binding" "analytics_write" {
+  project = var.gcp_project_id
+  role    = google_project_iam_custom_role.analytics_write_role.id
+
+  members = [
+    google_service_account.analytics_write.member
+  ]
+}
+
+###
+
+# service account for reading ga4 data
+resource "google_service_account" "analytics_read" {
+  account_id   = "ga4-read-bq"
+  display_name = "ga4-read-bq"
+  project      = "GA4-Analytics"
+}
+
+# custom role for analytics read
+resource "google_project_iam_custom_role" "analytics_read_role" {
+  role_id     = "analytics_read_role"
+  title       = "ga4-read-bq-permissions"
+  description = "Read source ga4 bq event data"
+
+  permissions = [
+    "bigquery.datasets.get",
+    "bigquery.tables.get",
+    "bigquery.tables.getData"
+  ]
+}
+
+# binding role to service account
+resource "google_project_iam_binding" "analytics_read" {
+  project = "GA4-Analytics"
+  role    = google_project_iam_custom_role.analytics_read_role.id
+
+  members = [
+    google_service_account.analytics_read.member
+  ]
+}
+
+###
+# top level dataset to store events for ingestion into vertex
+resource "google_bigquery_dataset" "dataset" {
+  dataset_id = "analytics_events_vertex"
+  project    = var.gcp_project_id
+}
+
+# ga4 'view_item_list' events get transformed and inserted into this time-partitioned search-event table defined with a vertex schema 
+resource "google_bigquery_table" "search-event" {
+  dataset_id          = google_bigquery_dataset.dataset.dataset_id
+  table_id            = "search-event"
+  project             = var.gcp_project_id
+  schema              = file("./files/search-event-schema.txt")
+  deletion_protection = true
+  time_partitioning {
+    field = "eventTime"
+    type  = "DAY"
+  }
+
+}
+
+# ga4 'select_item' events get transformed and inserted into this time-partitioned search-event table defined with a vertex schema 
+resource "google_bigquery_table" "view-item-event" {
+  dataset_id          = google_bigquery_dataset.dataset.dataset_id
+  table_id            = "view-item-event"
+  project             = var.gcp_project_id
+  schema              = file("./files/view-item-event-schema.txt")
+  deletion_protection = false
+  time_partitioning {
+    field = "eventTime"
+    type  = "DAY"
+  }
+}
+
+# bucket for function .zip
+resource "google_storage_bucket" "storage_analytics_transfer_function" {
+  name     = "storage_analytics_transfer_function"
+  location = var.gcp_region
+}
+
+# zipped function into bucket
+resource "google_storage_bucket_object" "function_zipped" {
+  name   = "func.zip"
+  bucket = google_storage_bucket.storage_analytics_transfer_function.name
+  source = data.archive_file.init.output_path
+}
+
+# archive .py and requirements.txt to zip
+data "archive_file" "init" {
+  type        = "zip"
+  source_dir  = "${path.module}/files/function/"
+  output_path = "${path.module}/files/init.zip"
```

Comment:
Let's consistently use `path.module` or `./` (not fussed about which one).

Reply:
Also: I think it'd be nice to have a consistent file/folder naming for this.

Maybe the folder can be `files/storage_analytics_transfer_function`, and the zip file both here and in the `google_storage_bucket_object.function_zipped` resource could be `storage_analytics_transfer_function.zip`?